### ISSUE TITLE
Objects

### ISSUE THREAD

**Issue by krakjoe** [krakjoe/parallel#32]
I've started to work on this, in a branch ... I've come up with a way to copy objects that works safely for pure user objects.

I was going to try to introduce support for internal, or internal derived objects, however, for this to work properly, we need an additional internal handler for objects; Should parallel ever be merged, we can talk about adding that and implementing the necessary handlers for all internal objects.

We could, theoretically abuse the clone object handler, however, some objects do not implement this properly now, and none of them are really prepared for it to be executed as we need to execute it, so it would probably just introduce more crashes than anything.

Any thoughts are welcome ...

**Comment by krakjoe** [comment]
This morning I'm not even sure I like this approach ...

Some questions ...

How important is object support to everyone else ? 
If it is important, what kind of support do you want for user objects ? 
Should we leave object support for later, when we may be able to introduce superior support in core ?

**Comment by dktapps** [comment]
I think that supporting user object without supporting pure internal objects is going to be a confusing inconsistency that should be avoided.

parallel is shaping up to look good so far, and it would be a bad idea imho if developers were forced to use strange hacks to get inconsistent behaviour like this to work the way they desire.

**Comment by dktapps** [comment]
As far as shared objects go, I see two approaches:
1) copy everything (why not just pass a serialized object to your closure at this point?)
2) The devil of pthreads `Threaded` objects (only these are true shared, everything else must then use approach (1) :/)

Neither of these approaches are without cognitive pain, so I think it's best left alone until it's possible to provide a fully consistent behaviour for everything.

**Comment by krakjoe** [comment]
The more I think about it, the less I like it ... I think we probably will just wait for more support from core. An unbuffered channel at the moment is the fastest it has ever been possible to share data between threads, or processes, and objects can't get close to that until we have more core support ...

**Comment by WyriHaximus** [comment]
@krakjoe Honestly don't add it if it doesn't make sense, or would compromise the integrity/stability of this extension IMHO

**Comment by krakjoe** [comment]
I'll just lay out my initial justification for doing the pure user object support ...

You can send a shape over a channel or back from a task with an array, however, it has to be copied as an array, key's, values, and buckets all add to the overhead of that operation. A pure user object has no keys (because they are contained in class definition, and ever present if (pre)loaded), it's extremely fast to do a bitwise copy of an object (no serialization), with no buckets and no keys where the properties are declared, and only one hash where they are not.

We can safely do this for pure user objects if we give them by-value semantics (which sort of makes sense for threading anyway).

We can do it safely for a lot of internals objects given the mechanisms to do it, which could either be a part of parallel, or merged into core at some future time ...

It's not the case that it would compromise the stability of parallel to do this, but it would introduce, perhaps prematurely, by-value semantics for objects that we couldn't likely change in future should core get more support.

Now onto the question of whether it's likely that core will ever get the kind of support we would need for anything but by-value semantics for objects ... it actually isn't, in reality, we are talking about a different garbage collector for PHP, we are talking about different memory management implementation, and most importantly, we are talking about a different core and extension architecture - share nothing goes out the window ...

By-value objects is likely the farthest we can ever take object support, so we are basically faced with three options:

  - do nothing - no object support, ever
  - do pure user objects - which in my view would cover the 85% use case for an object
  - do full object support - this is probably overkill, there are probably objects that simply cannot be shared, it would require an audit of every extension in existence, here be dragons ...

So, they are my thoughts on this topic in full ... I don't know what is the best way forward, but I'm happy to do nothing, and I'm happy to hear everyone else's thoughts too ...

**Comment by dbalabka** [comment]
@krakjoe I want to share my thoughts. I might be wrong in some statements because of a lack of deep knowledge of  PHP internals.

IMO objects support is a necessary feature; otherwise, developers will start to use the standard `serialization()` and exchange values using strings to overcome such limitation. This definitely will hit performance.

 > A pure user object has no keys (because they are contained in class definition, and ever present if (pre)loaded), it's extremely fast to do a bitwise copy of an object (no serialization), with no buckets and no keys where the properties are declared, and only one hash where they are not.

It sounds great that it is possible to get better performance with "pure objects." IMO it would be good to have a fallback strategy for other types of objects with worse performance.  
BTW Is this working in the same way with dynamically defined object properties? 

I take a look at [Swoole Channel documentation](https://www.swoole.co.uk/docs/modules/swoole-channel) and seems they always use serialization strategy for anything except strings. Probably to overcome performance issues they use their serialization implementation: https://www.swoole.co.uk/docs/modules/swoole-serialize

**Comment by krakjoe** [comment]
So, I thought long and hard about this, I spent some time reading the opinions of people who actually use go, because while I'm a little familiar with go, I do not actually use it.

The basic question is this: What kind of passing semantics do we want for variables in parallel ?

In go, everything is passed by value (although the meaning of the term value is not quite the same as it is in PHP), for go this is an opimization, they have mechanisms that mean that passing by value is often more efficient than passing by reference (pointer). While the people who wrote the compiler want it to work that way because they can optimize it, the people who use go want it to work that way because it reduces the cognitive overhead of parallelism. There are a couple of mantras thrown around in the go community:

  - Do not communicate by sharing memory; instead, share memory by communicating.
  - Data should be owned by one task

Conventionally, parallel code shares memory, and it's this that makes writing parallel code challenging, it's hard to model in your head and get everything right precisely because everything is operating in the same address space, all objects communicate their state by virtue of the fact that they are in the same address space, and an object is free to read the memory of another object at will.

In PHP, and in Go, tasks/routines are isolated from one another, they cannot (by default) share memory. They reduce the points in your application at which memory may be exchanged or read from another task to very well understood managed synchronization points - channels/events/select and so on. In go, it's possible to break this rule, you may pass pointers, and go exposes low level mutex to manage those pointers.

In PHP, it is not possible to break this rule safely because of the architecture of PHP - there is no sharing memory at the level of user code (memory is shared internally where possible, don't panic, we're not being wasteful). So we're forced into embracing this idea that you should share memory by communicating and not the inverse, and will never need to expose a low level mutex.

So the answer to the question, "what kind of passing semantics do we want?", is actually "pass by value, all the time".

An object in PHP is a zend_object, a zend_object may be copied by value very efficiently, as I mentioned before. An internal object in PHP is a wrapper around a zend_object, which associates additional memory with the PHP object: This additional memory cannot be copied by value, there is no mechanics to do that. We can copy the additional memory, but the objects would behave strangely - even if we played internals tricks to intercept all property accesses and method calls, because the additional memory would be getting shared by more than one thread, surprising things would happen.

Internal objects simply don't fit with the model of parallel, nor do they fit with the philosophy I think we should be embracing. In go, these mantras actually have to be worked at, you have to put effort into making sure that one task owns data, and that you only share by communicating (pointers don't leak and so on). In PHP, you don't have to put effort into these things, that's just the way parallel code *must* work.

There will be some people that ignore or don't know about best practice, and they surely will decide it's a good idea for tasks to share large object graphs that include possible internal objects, and they surely will have to resort to serialization.

If the ideal (and reality) is that data is owned by one task, and that task doesn't just require some dependency value (like a value object for configuring the task for example), but requires the kind of graph that would include internal objects (data structures or otherwise shared functionality), then it's *not an independent task*, and the fact you are trying to parallelize it is questionable.

I'm sure you're all bored of me talking, so I'm going to leave it there ... but the tl;dr is, we got good user object support, and I think we need not go further than that.

Note: It's simple to say "we support user objects", and I don't know what language I'll use in the manual, but the reality is we support pure zend objects, and they can be internal - an Event is a pure zend object, there may be other such objects in the wild, but I don't have a list of them.

Note2: In addition to pure zend objects, channels may be shared because they already support that and are core to parallel, and this should reduce make/open boiler plate, I'm not going to add a conventional constructor to channel, we'll stick with the named one, less methods is better ...

**Comment by dktapps** [comment]
as far as the internal objects support goes ... I can think of cases where I'd like to be able to directly copy, such as when using SplFixedArray, any `Ds` types, etc ... in those cases where it's not possible to do a simple copy, would it be reasonable to do a serialize/unserialize a la pthreads? and just throw an exception (as we do now) for non-serializable objects?

**Comment by krakjoe** [comment]
I don't want to introduce reliance on serialize and unserialize in
parallel, they are slow, they are also the cause of the majority of
security bugs in PHP and I don't have the time or will power to audit the
code to make it safe.

Sharing data structures doesn't make a lot of sense with the CSP model, I
understand it may help you to port code, but it would be better to think
about the problem differently.

Cheers
Joe

**Comment by dktapps** [comment]
well... back to userland serializing then I suppose.

There's no telling in some complex systems when you might happen to have an internal object referenced somewhere (for example PM plugins can do god knows what), and then we're back to square one but with a broken object on the other side, instead of an error at the point of copy.
